build_test/lib/src/debug_test_builder.dart

```diff
@@ -5,45 +5,133 @@
 import 'dart:async';
 
 import 'package:build/build.dart';
+import 'package:glob/glob.dart';
 import 'package:html/dom.dart';
 import 'package:html/parser.dart';
 import 'package:path/path.dart' as p;
 
-class DebugTestBuilder extends Builder {
-  @override
-  final buildExtensions = const {
-    '_test.dart': const ['_test.debug.html']
-  };
+const _inputExtension = '_test.dart';
+const _outputExtension = '_test.debug.html';
 
-  @override
-  Future build(BuildStep buildStep) async {
-    var customHtmlId = buildStep.inputId.changeExtension('.html');
+/// Returns the (optional) user-provided HTML file to use as an input.
+///
+/// For example, for `test/foo_test.dart`, we look for `test/foo_test.html`.
+AssetId _customHtmlId(AssetId dartTest) {
+  return dartTest.changeExtension('.html');
+}
+
+/// Returns the builder-generated HTML file for browsers to navigate to.
+AssetId _debugHtmlId(AssetId dartTest) {
+  return dartTest.changeExtension('.debug.html');
+}
+
+/// Returns the JS script path for the browser for [dartTest].
+String _jsScriptPath(AssetId dartTest) {
+  return '${p.basename(dartTest.path)}.js';
+}
+
+/// Generates a `*.debug.html` for every file in `test/**/*_test.dart`.
+///
+/// This is normally used in order to use Chrome (or another browser's)
+/// debugging tools (such as setting breakpoints) while running a test suite.
+class DebugTestBuilder implements Builder {
+  /// Generates a `.debug.html` for the provided `._test.dart` file.
+  static Future<void> _generateDebugHtml(
+    BuildStep buildStep,
+    AssetId dartTest,
+  ) async {
+    final customHtmlId = _customHtmlId(dartTest);
+    final jsScriptPath = _jsScriptPath(buildStep.inputId);
     String debugHtml;
-    var jsScriptPath = '${p.basename(buildStep.inputId.path)}.js';
     if (await buildStep.canRead(customHtmlId)) {
-      var htmlContent = await buildStep.readAsString(customHtmlId);
-      var document = parse(htmlContent);
+      debugHtml = _replaceCustomHtml(
+        await buildStep.readAsString(customHtmlId),
+        jsScriptPath,
+      );
+    } else {
+      debugHtml = _createDebugHtml(jsScriptPath);
+    }
+    return buildStep.writeAsString(_debugHtmlId(dartTest), debugHtml);
+  }
 
-      // Replace the <link rel="x-dart-test" href="some_path"> tag with
-      var linkTag = document.querySelector('link[rel="x-dart-test"]');
-      var scriptTag = new Element.tag('script');
-      scriptTag.attributes['src'] = jsScriptPath;
-      linkTag.replaceWith(scriptTag);
+  /// Returns the HTML body of [customHtml] modified to work with this package.
```

Comment:
[nit] When I read this I interpret 'body' to mean `<body>`. Maybe 'content' instead?

Reply:
Done